File: algorithms/graph_algorithms/dijkstra/dijkstra.cpp
// Dijkstra's Algorithm: Shortest path in weighted graphs
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <limits>
#include <algorithm>

using namespace std;

struct State {
    int cost;
    string node;
    vector<string> path;
    
    bool operator>(const State& other) const {
        return cost > other.cost;
    }
};


pair<vector<string>, int> dijkstra_heap(
    const unordered_map<string, vector<pair<string, int>>>& graph,
    const string& start, const string& goal) {
    
    priority_queue<State, vector<State>, greater<State>> pq;
    unordered_set<string> visited;
    
    pq.push({0, start, {start}});
    
    while (!pq.empty()) {
        State current = pq.top();
        pq.pop();
        
        if (visited.count(current.node)) continue;
        if (current.node == goal) {
            return {current.path, current.cost};
        }
        
        visited.insert(current.node);
        
        auto it = graph.find(current.node);
        if (it != graph.end()) {
            for (const auto& [neighbor, weight] : it->second) {
                if (!visited.count(neighbor)) {
                    auto new_path = current.path;
                    new_path.push_back(neighbor);
                    pq.push({current.cost + weight, neighbor, new_path});
                }
            }
        }
    }
    
    return {{}, numeric_limits<int>::max()};
}


pair<vector<string>, int> dijkstra_array(
    const unordered_map<string, vector<pair<string, int>>>& graph,
    const string& start, const string& goal) {
    
    unordered_map<string, int> distances;
    unordered_map<string, string> previous;
    unordered_set<string> unvisited;
    
    for (const auto& [node, _] : graph) {
        distances[node] = numeric_limits<int>::max();
        unvisited.insert(node);
    }
    distances[start] = 0;
    
    while (!unvisited.empty()) {
        string current;
        int min_dist = numeric_limits<int>::max();
        
        for (const auto& node : unvisited) {
            if (distances[node] < min_dist) {
                min_dist = distances[node];
                current = node;
            }
        }
        
        if (min_dist == numeric_limits<int>::max()) break;
        
        if (current == goal) {
            vector<string> path;
            string node = current;
            while (!node.empty()) {
                path.insert(path.begin(), node);
                node = previous[node];
            }
            return {path, distances[goal]};
        }
        
        unvisited.erase(current);
        
        auto it = graph.find(current);
        if (it != graph.end()) {
            for (const auto& [neighbor, weight] : it->second) {
                int new_dist = distances[current] + weight;
                if (new_dist < distances[neighbor]) {
                    distances[neighbor] = new_dist;
                    previous[neighbor] = current;
                }
            }
        }
    }
    
    return {{}, numeric_limits<int>::max()};
}


int main() {
    unordered_map<string, vector<pair<string, int>>> graph = {
        {"A", {{"B", 1}, {"C", 4}}},
        {"B", {{"C", 2}, {"D", 5}}},
        {"C", {{"D", 1}}},
        {"D", {}}
    };
    
    auto [path, cost] = dijkstra_heap(graph, "A", "D");
    for (const auto& node : path) cout << node << " ";
    cout << "with cost " << cost << endl;
    
    return 0;
}

File: algorithms/graph_algorithms/dijkstra/dijkstra.py
# Dijkstra's Algorithm: Shortest path in weighted graphs
from heapq import heappush, heappop


def dijkstra_heap(graph, start, goal):
    """Dijkstra with min-heap - O((V + E) log V)."""
    pq = [(0, start, [start])]
    visited = set()
    
    while pq:
        cost, node, path = heappop(pq)
        
        if node in visited:
            continue
        
        if node == goal:
            return path, cost
        
        visited.add(node)
        for neighbor, weight in graph.get(node, []):
            if neighbor not in visited:
                heappush(pq, (cost + weight, neighbor, path + [neighbor]))
    
    return None, float('inf')


def dijkstra_array(graph, start, goal):
    """Dijkstra with array - O(V²), simple but slower."""
    distances = {node: float('inf') for node in graph}
    distances[start] = 0
    previous = {}
    unvisited = set(graph.keys())
    
    while unvisited:
        # Find min distance node
        current = min(unvisited, key=lambda n: distances[n])
        
        if distances[current] == float('inf'):
            break
        
        if current == goal:
            path = []
            while current:
                path.insert(0, current)
                current = previous.get(current)
            return path, distances[goal]
        
        unvisited.remove(current)
        
        for neighbor, weight in graph.get(current, []):
            new_dist = distances[current] + weight
            if new_dist < distances[neighbor]:
                distances[neighbor] = new_dist
                previous[neighbor] = current
    
    return None, float('inf')


def dijkstra_all_paths(graph, start):
    """Dijkstra from single source to all nodes."""
    distances = {node: float('inf') for node in graph}
    distances[start] = 0
    pq = [(0, start)]
    visited = set()
    
    while pq:
        cost, node = heappop(pq)
        
        if node in visited:
            continue
        
        visited.add(node)
        
        for neighbor, weight in graph.get(node, []):
            new_dist = cost + weight
            if new_dist < distances[neighbor]:
                distances[neighbor] = new_dist
                heappush(pq, (new_dist, neighbor))
    
    return distances

